### PR TITLE
Add category to search

### LINKED ISSUE #1377: Allow categories to be searched
Allow search to look for categories. Supports #1183.

### PR DESCRIPTION
**Pull request by tofumatt**
Okay, this patch extends search to check through a list of allowed "filters" (right now it's app, extension type, query, and category). It will allow me to add a category page separately as we can now search via category.

I would say that, as @mstriemer chatted about on IRC, `SearchResults` should really just be `AddonList` or something like that, but as that's just semantics and I wanted to keep this as short as possible I left it for now.

This is ready for review though I think I mistakenly included a `CategoryPage` container I didn't need to; will remove in a few.

Fixes #1377

### CODE REVIEW

File: src/core/constants.js [hunk 1]
```diff
@@ -56,7 +56,7 @@ export const SET_CLIENT_APP = 'SET_CLIENT_APP';
 export const SET_CURRENT_USER = 'SET_CURRENT_USER';
 export const SET_JWT = 'SET_JWT';
 export const SET_LANG = 'SET_LANG';
-export const SET_QUERY = 'SET_QUERY';
+export const SET_FILTERS = 'SET_FILTERS';
```
Comment (tofumatt) [inline src/core/constants.js]: I'm actually not sure this is being used anywhere.

**Comment by tofumatt** [comment]
Alright, looks like this is ready for r?

I'll rebase when review is finished. 😄

**Comment by tofumatt** [comment]
Yeah this isn't actually being used yet, but I figured it was worth landing
this first to make the PR smaller. Like you said, the category page will
use it, should be a much smaller PR :-)

- tofumatt

File: src/core/searchUtils.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+export function mapStateToProps(state, ownProps) {
+  const { location } = ownProps;
+
+  const filters = makeFilters({
+    ...location.query,
+    clientApp: state.api.clientApp,
+  });
+  const hasSearchParams = Object.values(filters).some((param) => (
```
Comment (tofumatt) [inline src/core/searchUtils.js]: Oh, good point!

File: src/core/searchUtils.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+
+export const paramsToFilter = {
+  app: 'clientApp',
+  category: 'category',
+  q: 'query',
+  type: 'addonType',
+};
+
+export function makeFilters(params) {
```
Comment (tofumatt) [inline src/core/searchUtils.js]: Yeah, they convert query strings (which we try to match to AMO's existing ones as best we can) to the internal identifiers we use (to have nicer variable names) and then back again to the params the AMO API expects (eg. we use `clientApp` internally but the API expects `app`.

I'll add some comments explaining this but also will see about a more descriptive name.

File: src/core/searchUtils.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+export function isLoaded({ page, state, filters }) {
+  const stateMatchesLocation = Object.keys(filters).every((key) => (
+    filters[key] === state.filters[key]
+  ));
+  return stateMatchesLocation && state.page === page && !state.loading;
+}
+
+export function parsePage(page) {
+  const parsed = parseInt(page, 10);
+  return Number.isNaN(parsed) || parsed < 1 ? 1 : parsed;
```
Comment (tofumatt) [inline src/core/searchUtils.js]: Thanks @Ryuno-Ki, I've filed an issue for that page number parsing, if you'd like to work on it let me know, should actually be a good first bug for the repo. 😄

Comment (Ryuno-Ki) [inline src/core/searchUtils.js]: I've tried React on Node Knockout Hackathon last weekend. I didn't managed to get something basic as sending a XHR working … I should take more courses on CodeCademy before I get it … but they have a bug preventing me from progressing :(

**Pull request merged**